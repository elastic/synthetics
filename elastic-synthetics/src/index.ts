require('source-map-support').install();

export * from './dsl';
export * from './runner';

import { run } from './runner' ;

if (require.main === module)  {
    run()
}
