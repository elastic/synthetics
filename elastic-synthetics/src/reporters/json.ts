import BaseReporter from './base';

interface JourneyResults {
    id: string,
    name: string,
    meta: {[key: string]: any},
    elapsed_ms: number,
    steps: Array<{
        name: string,
        source: string,
        elapsed_ms: number,
        error: Error,
        screenshot: string,
    }>
}

export default class JSONReporter extends BaseReporter {
    _registerListeners() {
        const journeyMap = new Map<String, JourneyResults>();
        this.runner.on(
            'journeyStart',
            ({journey, params}) => {
                const { id, name } = journey.options;
                if (!journeyMap.has(name)) {
                    journeyMap.set(name, {
                        id,
                        name,
                        elapsed_ms: -1, // gets set at the end
                        meta: params,
                        steps: []
                    });
                }
            }
        );

        this.runner.on('journeyEnd', ({journey, elapsedMs }) => {
            journeyMap.get(journey.options.name).elapsed_ms = elapsedMs;
        })

        this.runner.on('stepEnd', ({journey, step, elapsedMs, error, screenshot}) => {
            const journeyOutput = journeyMap.get(journey.options.name);
            journeyOutput &&
                journeyOutput.steps.push({
                    name: step.name,
                    source: step.callback.toString(),
                    elapsed_ms: elapsedMs,
                    error,
                    screenshot
                });
        });

        this.runner.on('end', () => {
            this.write(this._getOutput(journeyMap));
        });
    }

    _getOutput(journeyMap) {
        const output = { journeys: [] };
        for (const journey of journeyMap.values()) {
            output.journeys.push(journey);
        }
        return output;
    }
}
