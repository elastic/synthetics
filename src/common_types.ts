import { Protocol } from 'playwright/types/protocol';

export type StatusValue = 'succeeded' | 'failed' | 'skipped';

export type FilmStrip = {
  snapshot: string;
  name: string;
  ts: number;
};

export type NetworkInfo = {
  url: string;
  method: string;
  type: string;
  request: Protocol.Network.Request;
  response: Protocol.Network.Response;
  isNavigationRequest: boolean;
  start: number;
  end: number;
  status: number;
};
