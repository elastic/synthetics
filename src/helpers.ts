import { red, green, yellow, cyan } from 'kleur/colors';
import os from 'os';
import path from 'path';
import { performance } from 'perf_hooks';

export function noop() {}

export function indent(lines: string, tab = '   ') {
  return lines.replace(/^/gm, tab);
}

export const symbols = {
  warning: yellow('⚠'),
  skipped: cyan('-'),
  succeeded: green('✓'),
  failed: red('✖'),
};

export function formatError(error: Error) {
  if (!(error instanceof Error)) {
    return;
  }
  const { name, message, stack } = error;
  return { name, message, stack };
}

export function generateTempPath() {
  return path.join(os.tmpdir(), `synthetics-${process.hrtime().toString()}`);
}

/**
 * As per the timings used in the Network Events from
 * Chrome devtools protocol
 */
export function getMonotonicTime() {
  const hrTime = process.hrtime();
  return hrTime[0] * 1 + hrTime[1] / 1e9;
}

/**
 * Timestamp at which the current node process began.
 */
const processStart = performance.timeOrigin;
export function getTimestamp() {
  return (processStart + now()) * 1000;
}

/**
 * Relative current time from the start of the current node process
 */
export function now() {
  return performance.now();
}

/**
 * Execute all the callbacks in parallel using Promise.all
 */
export async function runParallel(callbacks) {
  const promises = callbacks.map(cb => cb());
  return await Promise.all(promises);
}
