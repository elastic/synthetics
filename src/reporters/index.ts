/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import BaseReporter from './base';
import JSONReporter from './json';
import JUnitReporter from './junit';
import { Journey, Step } from '../dsl';
import {
  StartEvent,
  JourneyEndResult,
  JourneyStartResult,
  StepEndResult,
} from '../common_types';

export type ReporterOptions = {
  fd?: number;
  colors?: boolean;
  dryRun?: boolean;
};
export type BuiltInReporterName = 'default' | 'json' | 'junit';
export type ReporterInstance = new (opts: ReporterOptions) => Reporter;
export const reporters: {
  [key in BuiltInReporterName]: ReporterInstance;
} = {
  default: BaseReporter,
  json: JSONReporter,
  junit: JUnitReporter,
};

export interface Reporter {
  onStart?(params: StartEvent): void;
  onJourneyRegister?(journey: Journey): void;
  onJourneyStart?(journey: Journey, result: JourneyStartResult): void;
  onStepStart?(journey: Journey, step: Step): void;
  onStepEnd?(journey: Journey, step: Step, result: StepEndResult): void;
  onJourneyEnd?(
    journey: Journey,
    result: JourneyEndResult
  ): void | Promise<void>;
  onEnd?(): void | Promise<void>;
}
