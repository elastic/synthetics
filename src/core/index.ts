/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { Journey, JourneyCallback, JourneyOptions } from '../dsl';
import Runner from './runner';
import { VoidCallback } from '../common_types';
import { log } from './logger';

export const runner = new Runner();

export const journey = (
  options: JourneyOptions | string,
  callback: JourneyCallback
) => {
  log(`register journey: ${options}`)
  if (typeof options === 'string') {
    options = { name: options, id: options };
  }
  const j = new Journey(options, callback);
  runner.addJourney(j);
  return j;
};

export const step = (name: string, callback: VoidCallback) => {
  log(`register step: ${name}`)
  return runner.currentJourney?.addStep(name, callback);
};

export const beforeAll = (callback: VoidCallback) => {
  runner.addHook('beforeAll', callback);
};

export const afterAll = (callback: VoidCallback) => {
  runner.addHook('afterAll', callback);
};

export const before = (callback: VoidCallback) => {
  if (!runner.currentJourney) {
    throw new Error('before is called outside of the journey context');
  }
  return runner.currentJourney.addHook('before', callback);
};

export const after = (callback: VoidCallback) => {
  if (!runner.currentJourney) {
    throw new Error('after is called outside of the journey context');
  }
  return runner.currentJourney.addHook('after', callback);
};
