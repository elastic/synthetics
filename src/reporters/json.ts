import BaseReporter from './base';
import { StatusValue } from '../common_types';
import { formatError } from '../helpers';

// Semver version for the JSON emitted from this package.
const jsonFormatVersion = '1.0.0';

interface JourneyResults {
  id: string;
  name: string;
  meta: { [key: string]: any };
  duration_ms: number;
  url?: string; // URL at end of first step
  error?: Error;
  status: StatusValue;
  steps: Array<{
    name: string;
    source: string;
    duration_ms: number;
    error: Error;
    screenshot: string;
    status: StatusValue;
  }>;
}

export default class JSONReporter extends BaseReporter {
  _registerListeners() {
    const journeyMap = new Map<string, JourneyResults>();
    this.runner.on('journey:start', ({ journey, params }) => {
      const { id, name } = journey.options;
      if (!journeyMap.has(name)) {
        journeyMap.set(name, {
          id,
          name,
          meta: params,
          steps: [],
          duration_ms: 0,
          status: 'succeeded'
        });
      }
    });

    this.runner.on('journey:end', ({ journey, durationMs }) => {
      const journeyOutput = journeyMap.get(journey.options.name);
      journeyOutput.duration_ms = durationMs;
    });

    this.runner.on(
      'step:end',
      ({ journey, step, durationMs, error, screenshot, url, status }) => {
        const journeyOutput = journeyMap.get(journey.options.name);

        // The URL of the journey is the first URL we see
        if (!journeyOutput.url) {
          journeyOutput.url = url;
        }

        // If there's an error we hoist it up to the journey for convenience
        // and set the status to down
        if (error) {
          journeyOutput.error = formatError(error);
          journeyOutput.status = 'failed';
        }

        journeyOutput &&
          journeyOutput.steps.push({
            name: step.name,
            source: step.callback.toString(),
            duration_ms: durationMs,
            error: formatError(error),
            screenshot,
            status
          });
      }
    );

    this.runner.on('end', () => {
      this.write(this._getOutput(journeyMap));
    });
  }

  _getOutput(journeyMap) {
    const output = {
      __type__: 'synthetics-summary-results',
      format_version: jsonFormatVersion,
      journeys: []
    };
    for (const journey of journeyMap.values()) {
      output.journeys.push(journey);
    }
    return output;
  }
}
