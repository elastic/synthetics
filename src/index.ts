/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import { runner } from './core';
import { RunOptions } from './core/runner';
import { setLogger } from './core/logger';
import { parseArgs } from './parse_args';

/* eslint-disable @typescript-eslint/no-var-requires */
require('source-map-support').install();

export async function run(options: RunOptions) {
  const cliArgs = parseArgs();
  /**
   * Use the NODE_ENV variable to control the environment if its not explicity
   * passed from either CLI or through the API
   */
  options.environment = options.environment || process.env['NODE_ENV'];
  /**
   * set up logger with appropriate file descriptor
   * to capture all the DEBUG logs when running from heartbeat
   */
  const outfd = options.outfd ?? cliArgs.outfd;
  setLogger(outfd);

  try {
    return await runner.run({
      ...options,
      headless: options.headless ?? cliArgs.headless,
      screenshots: options.screenshots ?? cliArgs.screenshots,
      dryRun: options.dryRun ?? cliArgs.dryRun,
      journeyName: options.journeyName ?? cliArgs.journeyName,
      network: options.network ?? cliArgs.network,
      pauseOnError: options.pauseOnError ?? cliArgs.pauseOnError,
      reporter: cliArgs.json && !options.reporter ? 'json' : options.reporter,
      outfd,
    });
  } catch (e) {
    console.error('Failed to run the test', e);
    process.exit(1);
  }
}

export { beforeAll, afterAll, journey, step, before, after } from './core';
