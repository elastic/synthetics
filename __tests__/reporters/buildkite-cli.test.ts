/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

import * as jsonReporter from '../../src/reporters/json';

/**
 * Disable the ANSI codes for kleur/colors module
 */
process.env.NO_COLOR = '1';

import fs from 'fs';
import SonicBoom from 'sonic-boom';
import { step, journey } from '../../src/core';
import BuildKiteCLIReporter, {
  writeScreenshotsToPath,
} from '../../src/reporters/build_kite_cli';
import * as helpers from '../../src/helpers';
import { Screenshot } from '../../src/common_types';

describe('buildkite cli reporter', () => {
  let dest: string;
  let stream: SonicBoom;
  let reporter: BuildKiteCLIReporter;
  const timestamp = 1600300800000000;
  const j1 = journey('j1', () => {});

  const readAndCloseStream = async () => {
    /**
     * Close the underlying stream writing to FD to read all the contents
     */
    stream.end();
    await new Promise(resolve => stream.once('finish', resolve));
    return fs.readFileSync(fs.openSync(dest, 'r'), 'utf-8');
  };

  beforeEach(() => {
    dest = helpers.generateTempPath();
    reporter = new BuildKiteCLIReporter({ fd: fs.openSync(dest, 'w') });
    stream = reporter.stream;
    jest.spyOn(helpers, 'now').mockImplementation(() => 0);
  });

  afterEach(() => {
    fs.unlinkSync(dest);
  });

  afterAll(() => {
    process.env.NO_COLOR = '';
  });

  it('writes each step to the FD', async () => {
    reporter.onJourneyStart(j1, {
      timestamp,
    });
    reporter.onStepEnd(j1, step('s1', helpers.noop), {
      status: 'failed',
      error: {
        name: 'Error',
        message: 'step failed',
        stack: 'Error: step failed',
      },
      url: 'dummy',
      start: 0,
      end: 1,
    });
    reporter.onEnd();
    expect((await readAndCloseStream()).toString()).toMatchSnapshot();
  });

  it('render hook errors without steps', async () => {
    reporter.onJourneyStart(j1, {
      timestamp,
    });
    const error = {
      name: 'Error',
      message: 'before hook failed',
      stack: 'Error: before hook failed',
    };
    reporter.onJourneyEnd(j1, {
      timestamp,
      status: 'failed',
      error,
      start: 0,
      end: 1,
      options: {
        outputDir: 'screenshots',
      },
    });
    reporter.onEnd();
    expect((await readAndCloseStream()).toString()).toMatchSnapshot();
  });

  it('store screenshots in provided output dir', async () => {
    const data = {
      step: {
        name: 'do something',
        index: 2,
        location: {
          file: '/Users/shahzad/elastic/synthetics/test-projects/journeys/example.journey.ts',
          line: 12,
          column: 30,
        },
      },
      timestamp: 1693471547477914.5,
      data: '/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkS',
    } as Screenshot;

    jest
      .spyOn(jsonReporter, 'gatherScreenshots')
      .mockImplementation(
        async (
          screenshotsPath: string,
          callback: (data: Screenshot) => Promise<void>
        ) => {
          await callback(data);
        }
      );

    const spy = jest.spyOn(fs, 'writeFileSync');
    await writeScreenshotsToPath('on', 'screenshots', 'failed');
    expect(spy).toHaveBeenCalledWith(
      'screenshots/do something.jpg',
      '/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkS',
      { encoding: 'base64' }
    );
  });
});
