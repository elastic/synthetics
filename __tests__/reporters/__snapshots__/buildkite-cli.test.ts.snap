// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildkite cli reporter render hook errors without steps 1`] = `
"
--- Journey: j1
   
   thrown: {
     name: 'Error',
     message: 'before hook failed',
     stack: 'Error: before hook failed'
   }

✖ Took (1 seconds)
No tests found! (0 ms) 

"
`;

exports[`buildkite cli reporter writes each step to the FD 1`] = `
"
--- Journey: j1
   ✖  Step: 's1' failed (1000 ms)
   
   thrown: { name: 'Error', message: 'step failed', stack: 'Error: step failed' }

✖ Took (1 seconds)

 1 failed (0 seconds) 

"
`;

exports[`buildkite cli reporter writes multiple steps to the FD 1`] = `
"
--- Journey: j1
   ✓  Step: 's1' succeeded (1000 ms)
   ✓  Step: 's1' succeeded (4000 ms)
✓ Took (2 seconds)

 2 passed (0 seconds) 

"
`;
